### PR TITLE
config: No text/template wrapper

### PR DESCRIPTION
**Pull request by kellerza**
Nr 1 from #487 

Implementation similar to the existing logic from https://github.com/kellerza/template/blob/main/template.go#L59

@karimra mentioned there is a built in method to search the paths, but did not find it.

### CODE REVIEW

**Comment by karimra** [comment]
you can do something like this:

```go
func (c *CLab) parseTemplates(pathToTemplates string) (*template.Template, error) {
    t := template.New(c.Config.Name)
    err := filepath.Walk(pathToTemplates, func(path string, info os.FileInfo, err error) error {
        if strings.HasSuffix(path, ".gotmpl") { // or any other condition identifying the template files
            _, err = t.ParseFiles(path)
           return err
    })

    return t, err
}
```

The individual templates can be called with their name, which defaults to the file base name.

```go
t.ExecuteTemplate(os.Stdout, name, vars)
```

you can get the list of templates associated with `t`:
```go
t.Templates()
```

you can lookup a specific template associated with `t`:
```go
t.Lookup(name)
```

File: clab/config/template.go [hunk 1]
```diff
@@ -58,11 +75,21 @@ func RenderAll(nodes map[string]nodes.Node, links map[int]*types.Link) (map[stri
...
+
+			if tmpl.Lookup(tmplN) == nil {
+				err := LoadTemplate(tmpl, tmplN)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			var buf strings.Builder
+			err := tmpl.Lookup(tmplN).Execute(&buf, vars)
```
Comment (kellerza) [inline clab/config/template.go:87]:
```suggestion
			err := tmpl.ExecutTemplate(&buf, tmplN, vars)
```

**Comment by kellerza** [comment]
Thanks @karimra. Changed to `ExecuteTemplate`

Not sure walking the root path is applicable. The idea here is to look for a specific file if you have a list of possible paths.

File: clab/config/template.go [hunk 2]
```diff
@@ -58,11 +75,21 @@ func RenderAll(nodes map[string]nodes.Node, links map[int]*types.Link) (map[stri
...
+
+			if tmpl.Lookup(tmplN) == nil {
+				err := LoadTemplate(tmpl, tmplN)
+				if err != nil {
+					return nil, err
+				}
+			}
+
+			var buf strings.Builder
+			err := tmpl.ExecutTemplate(&buf, tmplN, vars)
```
Comment (kellerza) [inline clab/config/template.go:87]:
```suggestion
			err := tmpl.ExecuteTemplate(&buf, tmplN, vars)
```

**Comment by karimra** [comment]
It's still possible to parse the list of directories, and call the desired template by name.
I'm not sure about the logic you want to implement.

Parsing the whole directory, helps combining templates from different files as well.

**Comment by kellerza** [comment]
If we need templates from multiple files that is a requirement I did not have in mind. Do you see this as a requirement?

It seems quite ineffective, because of the possible amount of templates in a folder. You will end up loading all templates, for all kinds, over all folders specified.

If that is indeed the case, we should just run `ParseGlob` over all folders in order. From the docs:

> When parsing multiple files with the same name in different directories, the last one mentioned will be the one that results.  

Maybe with globs we could keep some of the existing logic, but only load all templates for the active role/kind

**Comment by karimra** [comment]
Templates can be combined by importing each other. In that case you would need to load the main template as well as the ones it imports.

Separating templates in multiples files allows reuse and separation of responsibilities/concerns, it's really at the core of the template engine design.

I don't think this has to be addressed in this PR, since we didn't really come to a conclusion concerning how templates will be mapped to kinds and nodes.

Can explain how a template file is mapped to a node in this PR? Is it a file name that gets inherited from kinds/defaults like the other variables?

**Comment by kellerza** [comment]
You can pass a list of templates names. The template names will be combined with the role: `<tname>__<role>.tmpl`

By default the role is the kind, except if you specify it in vars.

**Comment by hellt** [comment]
> You can pass a list of templates names. The template names will be combined with the role: `<tname>__<role>.tmpl`
> 
> By default the role is the kind, except if you specify it in vars.

yes, the idea here is that the template name is not related to a single node, but to a group of them. The default group marker is its `kind`.

So for a lab that has `srl` and `sros` kinds, this will allow to pass a template named `interfaces` files `interfaces__srl.tmpl` and `interfaces__sros.tmpl` will be looked up and applied to the relevant nodes

**Comment by kellerza** [comment]
I would phrase the *group of them* slightly differently:

- The template <name> gets applied to *all* nodes in the topology file.
- The role/kind decides which template (SRL syntax,SROS syntax, etc) get rendered for that node.

Any finer control can be done inside the template with if constructs using even more of the vars on the node/link level.

**Comment by kellerza** [comment]
I would add that the `role` variable creates groups: 
1. firstly each node kind will have its own CLI syntax (default=kind), but
2. nodes can also be grouped into roles (like `spine` or `leaf` roles, or even `leaf_srl` and `leaf_sros`)

**Comment by karimra** [comment]
it seems all this assumes that templates are being run with separate command, not run on deploy ?
can you show a topology file example where the templates are specified and the command to apply them ?

**Comment by kellerza** [comment]
Indeed, config only deploys to running labs.
Vr-05 example folder has 2 examples and the base templates are in the active branch

try: 

`clab config -t topo.yml -n base -p ../templates/config`

add -c 100 to view the rendered templates if you dont have an active lab

**Comment by karimra** [comment]
ok, if this is the case currently, `ParseGlob` works for now. I guess the glob can be adjusted to the pattern that matches the node name/role

**Comment by hellt** [comment]
@karimra do you have anything else you had in mind for this PR?
otherwise we can merge it

**Comment by karimra** [comment]
it's still hard for me to see the whole picture, go ahead. once the other PR is done it will hopefully be clearer.

**Comment by kellerza** [comment]
I'll make some slides to capture the whole picture, I've been asked to present it to an internal community as well. 

This should also give a start for the site/docs.

**Comment by karimra** [comment]
I get how it works and how to use it, what I don't get is the Go code. 
There seems to be a lot of assumptions on the template format and nature.

**Pull request merged**